metadata-ingestion/src/datahub/ingestion/source/cassandra.py

```diff
@@ -0,0 +1,413 @@
+)
+from datahub.emitter.mce_builder import (
+    make_data_platform_urn,
+    make_dataplatform_instance_urn,
+    make_dataset_urn_with_platform_instance,
+)
+from datahub.metadata.schema_classes import UsageAggregationClass, DataPlatformInstanceClass
+from cassandra.cluster import Cluster, Session
+from cassandra.auth import PlainTextAuthProvider
+from ssl import SSLContext, PROTOCOL_TLSv1_2, CERT_NONE
+from datahub.metadata.com.linkedin.pegasus2avro.common import StatusClass
+from datahub.metadata.schema_classes import (
+    ArrayTypeClass,
+    BooleanTypeClass,
+    BytesTypeClass,
+    DataPlatformInstanceClass,
+    DatasetProfileClass,
+    DatasetPropertiesClass,
+    DateTypeClass,
+    TimeTypeClass,
+    NullTypeClass,
+    NumberTypeClass,
+    OtherSchemaClass,
+    RecordTypeClass,
+    SchemaFieldDataTypeClass,
+    StringTypeClass,
+    SubTypesClass,
+)
+from datahub.metadata.com.linkedin.pegasus2avro.schema import (
+    SchemaField,
+    SchemaFieldDataType,
+    SchemaMetadata,
+)
+
+
+logger = logging.getLogger(__name__)
+
+
+# -------------------------------------------------- constants -------------------------------------------------- #
+PLATFORM_NAME_IN_DATAHUB = "cassandra"
+
+# we always skip over ingesting metadata about these keyspaces
+# TODO: make this configurable?
+SYSTEM_KEYSPACE_LIST = set(["system", "system_auth", "system_schema", "system_distributed", "system_traces"])
+
+# this keyspace contains details about the cassandra cluster's keyspaces, tables, and columns
+SYSTEM_SCHEMA_KESPACE_NAME = "system_schema"
+# these are the names of the tables we're interested in querying metadata from
+CASSANDRA_SYSTEM_SCHEMA_TABLES = {
+    "keyspaces": "keyspaces",
+    "tables": "tables",
+    "columns": "columns",
+}
+# these column names are present on the system_schema tables
+CASSANDRA_SYSTEM_SCHEMA_COLUMN_NAMES = {
+    "keyspace_name": "keyspace_name",
+    "table_name": "table_name",
+    "column_name": "column_name",
+    "column_type": "type",
+}
+
+# -------------------------------------------------- queries -------------------------------------------------- #
+# get all keyspaces
+GET_KEYSPACES_QUERY = f"SELECT * FROM {SYSTEM_SCHEMA_KESPACE_NAME}.{CASSANDRA_SYSTEM_SCHEMA_TABLES['keyspaces']}"
+# get all tables for a keyspace
+GET_TABLES_QUERY = f"SELECT * FROM {SYSTEM_SCHEMA_KESPACE_NAME}.{CASSANDRA_SYSTEM_SCHEMA_TABLES['tables']} WHERE {CASSANDRA_SYSTEM_SCHEMA_COLUMN_NAMES['keyspace_name']} = %s"
+# get all columns for a table
+GET_COLUMNS_QUERY = f"SELECT * FROM {SYSTEM_SCHEMA_KESPACE_NAME}.{CASSANDRA_SYSTEM_SCHEMA_TABLES['columns']} WHERE {CASSANDRA_SYSTEM_SCHEMA_COLUMN_NAMES['keyspace_name']} = %s AND {CASSANDRA_SYSTEM_SCHEMA_COLUMN_NAMES['table_name']} = %s"
+
+# -------------------------------------------------- setup the config and source reporter -------------------------------------------------- #
+
+# config
+class CassandraSourceConfig(PlatformInstanceConfigMixin, EnvConfigMixin):
```

Comment:
Wondering about the Stateful mixin. Might be easier to discuss over a short call if possible.